main/src/ca/uwaterloo/flix/language/phase/Namer.scala

```diff
@@ -1255,30 +1288,46 @@ object Namer extends Phase[WeededAst.Program, NamedAst.Root] {
       NamedAst.TypeParam(p, tvar, p.loc)
   }
 
+  // make sure all implicit names are defined explicitly, then use their types
+  // MATT docs
+  private def unifyTypeParams(explicitIdents: List[Name.Ident], implicitTypes: List[NamedAst.TypeParam])(implicit flix: Flix): Validation[List[NamedAst.TypeParam], NameError] = {
+    def checkDeclared(explicitIdents: Set[String], implicitType: NamedAst.TypeParam): Validation[NamedAst.TypeParam, NameError] = {
+      if (explicitIdents.contains(implicitType.name.name)) {
+        implicitType.toSuccess
+      } else {
+        NameError.UndefinedTypeVar(implicitType.name.name, implicitType.loc).toFailure
+      }
+    }
+    val explicitNames = explicitIdents.map(_.name).toSet
+    traverse(implicitTypes)(checkDeclared(explicitNames, _))
+  }
+
```

Comment:
I guess I would like to understand what is going on in the namer at a high-level.

We have to introduce fresh type variables and I assume the problem is that we do not always know their type? 

Could this be computed by looking through the type signatures or no?

What is the current strategy?

Reply:
We need to create a type variable for each, and when we create a type variable, we also declare its kind. But we cannot know the kind just from the explicit type parameter. The strategy used here was:
* Don't create a new type variable for each explicit type parameter
* Make sure all explicit type param names are used in the signature
  * (really should be the job of redundancy checker but since we don't create type variables for the explicit type params, this isn't accomplished there)
* Make sure all names used in the signature are among the explicit type parameters

> We have to introduce fresh type variables and I assume the problem is that we do not always know their type?

Assuming you mean "... know their **kind**": That's correct. We need to create a type variable for each, and when we create a type variable, we also declare its kind. But we cannot know the kind just from the explicit type parameter.

> Could this be computed by looking through the type signatures or no?

That's more or less what I'm attempting here.

> What is the current strategy?

The strategy used here was:
* Don't create a new type variable for each explicit type parameter
* Make sure all explicit type param names are used in the signature
  * (really should be the job of redundancy checker but since we don't create type variables for the explicit type params, this isn't accomplished there)
* Make sure all names used in the signature are among the explicit type parameters

All that said, I'm not very happy with the result of this strategy. After rebasing and some additional work, I decided on an alternative strategy which I mentioned briefly previously: unbound kinds, where we introduce a new kind to represent a sort of top kind which we assign to the explicit type params. Then during unification we ensure the substitution is built with `Unbound` kinds on the left so that we have proper kinds in the end. This seems to be a more elegant solution, but I have yet to fully explore the implications, especially with the rigidity changes. You can see the changes in a draft PR which I'll comment below after I've pushed it.